Drivers/Periphs/Src/CANBus.c

```diff
@@ -0,0 +1,189 @@
+ * @param mode CAN_MODE_NORMAL or CAN_MODE_LOOPBACK for operation mode
+ * @param queue (QueueHandle_t) initialized FreeRTOS queue for recieved messages
+ * @return HAL_StatusTypeDef - Status of CAN configuration
+ */
+HAL_StatusTypeDef CAN_Config(
+        CAN_HandleTypeDef *hcan,
+        uint32_t mode,
+        QueueHandle_t *queue) {
+    HAL_CAN_1 = hcan;
+    RxQueue = queue;
+    HAL_StatusTypeDef configstatus = MX_CAN1_Init(mode);
+    if (configstatus != HAL_OK) return configstatus;
+
+    CAN_FilterTypeDef filterconfig;
+    filterconfig.FilterBank = FILTER_BANK;
+    filterconfig.FilterMode = CAN_FILTERMODE_IDMASK;
+    filterconfig.FilterScale = CAN_FILTERSCALE_32BIT;
+    filterconfig.FilterIdLow = FILTER_ID_LOW;
+    filterconfig.FilterIdHigh = FILTER_ID_HIGH;
+    filterconfig.FilterMaskIdLow = FILTER_MASK_ID_LOW;
+    filterconfig.FilterMaskIdHigh = FILTER_MASK_ID_HIGH;
+    filterconfig.FilterFIFOAssignment = CAN_RX_FIFO_NUMBER;
+    filterconfig.FilterActivation = CAN_FILTER_ENABLE;
+    filterconfig.SlaveStartFilterBank = 14;
+
+    // Setup filter
+    configstatus = HAL_CAN_ConfigFilter(HAL_CAN_1, &filterconfig);
+    if (configstatus != HAL_OK) return configstatus;
+
+    // Start actual CAN
+    configstatus = HAL_CAN_Start(HAL_CAN_1);
+    if (configstatus != HAL_OK) return configstatus;
+
+    // Enable interrupt for pending rx message
+    configstatus = HAL_CAN_ActivateNotification(HAL_CAN_1, 
+        (CAN_RX_FIFO_NUMBER == CAN_RX_FIFO0) ? 
+            CAN_IT_RX_FIFO0_MSG_PENDING : CAN_IT_RX_FIFO1_MSG_PENDING);
+
+    return configstatus;
+}
+
+/** CAN Transmit Message
+ * @brief Transmit message over CAN
+ * @note This is really basic and does not check for a full transmit Mailbox
+ * 
+ * @param StdId Message ID (Standard)
+ * @param TxData Data to transmit
+ * @param len Length of data (Bytes) to transmit (MAX 8B)
+ * @return HAL_StatusTypeDef - Status of CAN configuration
+ */
+HAL_StatusTypeDef CAN_TransmitMessage(
+        uint32_t StdId,
+        uint8_t *TxData,
+        uint8_t len) {
+
+    CAN_TxHeaderTypeDef txheader;
+
+    // ExtID and extended mode are unused/not configured
+    txheader.StdId = StdId;
+    txheader.RTR = CAN_RTR_DATA;
+    txheader.IDE = CAN_ID_STD;
+    txheader.DLC = len;
+    txheader.TransmitGlobalTime = DISABLE;
+
+    return HAL_CAN_AddTxMessage(HAL_CAN_1, &txheader, TxData, &TxMailbox);
+}
+
+
+/**
+ * @brief Function executed by interrupt when there is a pending message
+ *        on RxFifo0
+ */
+void HAL_CAN_RxFifo0MsgPendingCallback(CAN_HandleTypeDef* hcan) {
+    HAL_CAN_GetRxMessage(hcan, CAN_RX_FIFO0, &RxHeader, RxData);
+    CAN_Recieve(&RxHeader, RxData);
+}
+
+/**
+ * @brief Function executed by interrupt when there is a pending message
+ *        on RxFifo1
+ */
+void HAL_CAN_RxFifo1MsgPendingCallback(CAN_HandleTypeDef* hcan) {
+    HAL_CAN_GetRxMessage(hcan, CAN_RX_FIFO1, &RxHeader, RxData);
+    CAN_Recieve(&RxHeader, RxData);
+}
```

Comment:
Why are there 2 FIFO's that can be used by CAN?

Reply:
Sunlight's CAN configuration sets up CAN_RX_FIFO0, but I left code in there supporting CAN_RX_FIFO1 in case it's ever needed. Should I remove all the CAN_RX_FIFO1 stuff?

Reply:
If we don't need it then I think it would be a good idea to get rid of it. We can include in the documentation that it is an option in case we need it.

Reply:
One other option would be to have just one function that reads from the right FIFO based on the config function.